Started to revise the pep (#8)
TobiasWrigstad/peps: pep-9999.rst

immutability status. Details about the default value of
this flag is discussed further down in this document. 

The flag can be added without extending the size of the
Python object header.

Implementation of Immutability
------------------------------

Immutability is enforced through run-time checking. The macro
``Py_CHECKWRITE(op)`` is inserted on all paths that are guaranteed
to end up in a write to ``op``. The macro inspects the immutability
flag in the header of ``op`` and signals an error if the immutability
flag is set.

A typical use of this check looks like this:

.. code-block:: c

  if (!Py_CHECKWRITE(op)) {        // perform the check
      PyErr_WriteToImmutable(op);  // raise the error if the check fails
      return NULL;                 // abort the write
  }  
  ... // code that performs the write

Writes are common in the CPython code base and the writes lack a
common "code path" that they pass. To this end, the PEP requires a
``Py_CHECKWRITE`` call to be inserted and there are several places
in the CPython code base that are changed as a consequence of this
PEP. So far we have identified around 70 writes spread across a

> dozen files. **TODO: update this number**

Author: @matajoh Can you update this paragraph?

easy to explain and understand, and an object's immutability can
be "trusted" --- it is not possible for an immutable object to
change through some nested mutable state [#RC]_. At the same time
it limits the utility of freezing as many Python objects contain
types outside of the standard library defined in C, which must
opt-in immutability before they can be frozen.

This PEP proposes immutability to be strict.

Dealing with Failure During Freezing
------------------------------------

Regardless whether support for freezing is opt-in or opt-out some
types will not be freezable. **TODO: give good examples**
This raises the question how to handle
failure to freeze an object graph. Consider the object graph
``o1 --> o2 --> o3`` where ``o1`` and ``o3`` can be frozen, but
``o2`` cannot. What are the possible behaviours of ``freeze(o1)``?

1. Freeze fails completely. In the strict interpretation of deep
   immutability, freezing ``o1`` is not possible because ``o1``
   contains a reference to an un-freezable object ``o2``. In this
   scenario, the object graph ``o1 --> o2 --> o3`` remains mutable
   and ``freeze(o1)`` raises an exception when the object graph

>    traversal encounters ``o2``.

Author: This is a great point! I am not sure that is as user-friendly, but it would promote more objects becoming immutable. Should we move this discussion to Teams?

succeeds or fails completely.

New Obligations on C Extensions
-------------------------------

Due to the opt-in decision, there are no *obligations* for C
extensions that does not want to add support for immutability. To
support immutability a class must **TODO: describe this**.
Depending on how state is kept and modified, additional changes
are needed as outlined below.

Because our implementation builds on information available to the CPython
cycle detector, types defined through C code will support immutability 
"out of the box" as long as they use Python standard types to store
data and uses the built-in functions of these types to modify the data.

To opt-in support for immutability, C extensions with functions
that directly write to data which can be made immutable should add
the ``Py_CHECKWRITE`` macro shown above on all paths in the code
that lead to writes to that data. Notably, if C extensions manage
their data through Python objects, no changes are needed.

Python objects may define a ``__freeze__`` method which will be
called **after** an object has been made immutable. This hook can
be used to freeze or otherwise manage any other state on the side

> that is introduced through a C-extension.

Author: Sorry, I don't understand what you mean.

race on the ``Counter`` class type object.

As types are frozen, this problem is avoided. Note that
freezing a class needs to freeze its superclasses as well.

Subclassing Immutable Classes
-----------------------------

CPython classes hold references to their subclasses. If
immutability it taken literally, it would not be permitted to
create a subclass of an immutable type. Because this reference
does not get exposed to the programmer in any dangerous way, we
permit frozen classes to be subclassed (by mutable classes). C.f.
`Sharing Immutable Data Across Subinterpreters`_.

Freezing Function Objects
-------------------------

Freezing function objects and lambdas is suprisingly involved
because all function objects have a pointer to ``globals``.
Function objects can be thought of as regular objects whose
fields are its local variables -- some of which may be captured
from enclosing scopes.

Consider the following scenario:

.. code-block:: python

  from freeze import freeze

  def example1():
      x = 0

      def foo():
          return x

      freeze(foo)
      ... # some code, e.g. pass foo to another thread
      x = 1
      foo()

>   example1()

Author: Just so that I am clear, will freezing `foo()` behave differently depending on whether `foo()` is in the global scope or a nested function? That does not feel great to me!

>   example1()

Reviewer: That is how Python works.  Global scoped functions capture globals by name, where as inside a function scope, they are boxed into PyCells.

>   example1()

Author: I understand that. But I am just second guessing our function freezing design that exposes this to users.

>   example1()

Author: Following discussion with @mjp41 -- we keep the above and aim to treat freezing functions uniformly regardless of the nature of the enclosing scope. I will also add a note saying that the prototype does not yet do this.

``foo`` to another thread will make the assignment ``x = 1`` a potential
data race.

We consider freezing of a function to freeze that function's
meaning at that point in time. In the code above, that means that
``foo`` gets its own copy of ``x`` which will have value of the enclosing
``x`` at the time of freezing, in this case 0.

Thus, the assignment ``x = 1`` is still permitted as it will not affect
``foo``, and it may therefore not contribute to a data race. Furthermore,
the result of calling ``foo()`` will be 0 -- not 1!

This is implemented by having ``x`` in ``foo`` point to a fresh
cell and then freezing the cell (and similar for global capture).
Note that this also prevents ``x`` from being reassigned.

We believe that this design is a sweet-spot that is intuitive and
permissive. (A **rejected alternative** is to freeze ``x`` in the
enclosing scope. This is problematic when a captured variable is
in ``globals()`` and also rejects more programs.)

Now consider freezing the following function:

.. code-block:: python

  from freeze import freeze

  def example2():
      x = 0
      def foo(a = False):
          if a:
              a = a + 1

>               return a

Author: The idea is to illustrate that mutating local variables work — I'd like to keep this line. Maybe we can just add a comment saying `# updating local variables work, even in a frozen function`. Would that work?

>               return a

Reviewer: Sure.  It needs explaining as without explanation it looks strange to me.

>               return a

Author: Yes, great.

>           if a:
>               a = a + 1
>               return a

Author: ```suggestion
          if a:
              a = a + 1 # Note: updating local variables work, even in a frozen function
              return a
```

   etc., instead of freezing existing objects. However, this does
   not generalize well to arbitrary objects and adds considerable
   complexity to all code bases.
4. Deep freezing immutable copies as proposed in `PEP 351: The 
   freeze protocol <https://peps.python.org/pep-0351/>`_. That PEP
   is the spiritual ancestor to this PEP which tackles the
   problems of the ancestor PEP and more (e.g. meaning of
   immutability when types are mutable, immortality, etc).
5. Deep freezing replaces data races with exceptions on attempts to
   mutate immutable objects. Another alternative would be to keep
   objects mutable and build a data-race detector that catches read--write
   and write--write races. This alternative was rejected for two main
   reasons:

    1. It is expensive to implement: it needs a read-barrier to
       detect what objects are being read by threads to capture
       read--write races.
    2. While more permissive, the model suffers from non-determinism.
       Data races can be hidden in corner cases that require complex
       logic and/or temporal interactions which can be hard to
       test and reproduce.

Open Issues
===========

> 1. How does deep freezing interact with weak references?

Author: Will you put on your TODO to write a short section on this?

> 1. How does deep freezing interact with weak references?

Reviewer: So I just hit a bug in the SCC prototype related to this.  Should a weak reference be converted to an effectively strong reference if it is frozen?

Consider:
```mermaid
stateDiagram
  r1 --> a
  a  --w--> b
  r2 --> b
```
If we `freeze(r1)`, and then `del(r2)`.  Should `a` remain reachable?

Consider:
```mermaid
stateDiagram
  r1 --> a
  a  --w--> b
  b  --> a
```

If we `freeze(a)`, should `a` and `b` be in the same SCC.

> 1. How does deep freezing interact with weak references?

Author: This is a super good point! I'll need to think about this.

objects are transmitted across subinterpreters.

This change will require reference counts to be atomic (as
discussed above) and the subclass list of a type object to
be made thread-safe. Additionally, we will need to change
the API for getting a class' subclasses in order to avoid
data races.

This change requires modules loaded in one subinterpreter to be
accessible from another. Implementation details here are to be
discussed, but the version we have been working on is one in which
a side effect of calling ``freezeglobals()`` is that all
subsequent module imports are imported into the main
module and immediately frozen and shared across all subinterpreters.

Data-Race Free Python
---------------------

While useful on their own, all the changes above are building
blocks of Data-Race Free Python. Data-Race Free Python will
borrow concepts from ownership (namely region-based ownership,
see e.g. `Cyclone <https://cyclone.thelanguage.org/>`_) to make Python programs data-race free
by construction. Which will permit multiple subinterpreters to
share *mutable* state, although only one subinterpreter at a time
will be able to access (read or write) to that state. In theory,

> this work could also be authored on-top of free-theaded Python (PEP 703).

Author: Yes, great point! Do we have a clear understanding of what we can put up as a preprint?

> this work could also be authored on-top of free-theaded Python (PEP 703).

Reviewer: I think the version we submitted as Camera Ready.

> this work could also be authored on-top of free-theaded Python (PEP 703).

Author: I'll upload that and link

It is important to point out that Data-Race Free Python is different
from `PEP 703`_, but aims to be fully compatible with that PEP, and
we believe that both PEPs can benefit from each other. In essence
`PEP 703`_'s focus is on making the CPython run-time resilient against
data races in Python programs: a poorly synchronized Python program
should not be able to corrupt reference counts, or other parts of
the Python interpreter. The complementary goal pursued by this PEP
is to make it impossible for Python programs to have data races.
Support for deeply immutable data is the first important step
towards this goal.

Reference Implementation
========================

`Available here <https://github.com/mjp41/cpython/pull/51>`_.

Rebasing on Python 3.14
=======================

> **TODO**

Author: Good point. A strawman would be to freeze only trees…

> **TODO**

Author: I was thinking that if the RC=1 (== tree) then there won't be any mutation under the hood.

> **TODO**

Reviewer: Ah, okay, that would work, but is pretty restrictive

> **TODO**

Author: note: we discussed locking and backing out on races (including freeze-freeze races…)

PEP: 9999
Title: Deep Immutability in Python
Author: Matthew Johnson <matjoh@microsoft.com>, Matthew Parkinson <mattpark@microsoft.com>, Sylvan Clebsch <sylvan.clebsch@microsoft.com>, Fridtjof Peer Stoldt <fridtjof.stoldt@it.uu.se>, Tobias Wrigstad <tobias.wrigstad@it.uu.se>

> Sponsor: TBD

Author: Who should be our sponsor? Does that happen *now* or later?

> Discussions-To: TBD

Author: Do we know how to open a discussion on discord?

Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 27-Feb-2025
Python-Version: 3.15
Post-History: 
Resolution:

Abstract
========

This PEP proposes adding a mechanism for deep immutability to
Python. The mechanism requires some changes to the core language,
but user-facing functions are delivered in a module called

> ``freeze``. This module provides two functions:

Author: Module called `immutable`

1. ``freeze(obj)`` -- turns ``obj`` deeply immutable
2. ``is_immutable(obj)`` -- returns ``True`` if ``obj`` is immutable

Making an object *deeply* immutable recursively renders the object
*and all objects it references* immutable. (Just
making the first object immutable is called *shallow*
immutability.)

Deep immutability provides strong guarantees against
unintended modifications, thereby improving correctness, security, and
parallel execution safety.

> Immutable objects are not immortal -- their reference counts

Author: not immortal by default // orthogonal

memory management extensions, including support for atomic
reference counting on immutable objects. These are outlined at the
end of this document.

Immutability in action:

.. code-block:: python

  from freeze import freeze, is_immutable

  class Foo:
      pass

  f = Foo()
  g = Foo()
  h = Foo()

  f.f = g
  g.f = h
  h.f = g # cycles are OK!
  del g # Remove local ref to g, so g's RC = 1
  del h # Remove local reg to h, so h's RC = 1

  g.x = "African Swallow" # OK
  freeze(f) # Makes, f, g and h immutable
  g.x = "European Swallow" # Throws an exception "g is immutable"
  is_immutable(h) # returns True
  h = None # Cycle detector will eventually find and collect the cycle

Motivation
==========

Ensuring Data Integrity
-----------------------

Python programs frequently manipulate large, interconnected data
structures such as dictionaries, lists, and user-defined objects.
Unintentional mutations can introduce subtle and
difficult-to-debug errors. By allowing developers to explicitly
freeze objects and their transitive dependencies, Python can
provide stronger correctness guarantees for data processing
pipelines, functional programming paradigms, and API boundaries

> where immutability is beneficial.

Reviewer 2: Should we maybe mention caching here? So the fact that objects returned from a cache are usually mutable which can lead to interesting results if they changed?

  stemming from this design.
* Global State Impact: Freezing an object that references global
  state (e.g., ``sys.modules``, built-ins) could inadvertently
  freeze critical parts of the interpreter. **TODO: still true?**
  **Mitigation:** Avoiding accidental freezing is possible by
  storing a non-freezable object in an object which should not be
  frozen. We will provide a singleton such object as part of the
  ``freeze`` module. Also, when the Python interpreter is exiting, we make
  all immutable objects mutable to facilitate a clean exit of the
  interpreter.

As the above list shows, a side-effect of freezing an object is
that its type becomes frozen too. Consider the following program,
which is not legal in this PEP because it modifies the type of an
immutable object:

.. code-block:: python

  from freeze import freeze

  class Counter:
      def __init__(self, initial_value):
          self.value = initial_value
      def inc(self):
          self.value += 1
      def dec(self):
          self.value -= 1
      def get(self):
          return self.value

  c = Counter(0)
  c.get() # returns 0 
  freeze(c) # (*) -- this locks the value of the counter to 0
  ... 

>   Counter.get = lambda self: 42 # OK in CPython 3.12, throws exception with this PEP

Author: Clarify why

PEP: 9999
Title: Deep Immutability in Python
Author: Matthew Johnson <matjoh@microsoft.com>, Matthew Parkinson <mattpark@microsoft.com>, Sylvan Clebsch <sylvan.clebsch@microsoft.com>, Fridtjof Peer Stoldt <fridtjof.stoldt@it.uu.se>, Tobias Wrigstad <tobias.wrigstad@it.uu.se>
Sponsor: TBD
Discussions-To: TBD
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 27-Feb-2025
Python-Version: 3.15
Post-History: 
Resolution:

Abstract
========

This PEP proposes adding a mechanism for deep immutability to
Python. The mechanism requires some changes to the core language,
but user-facing functions are delivered in a module called
``immutable``. This module provides the following functions and types:

1. The function ``freeze(obj)`` -- turns ``obj`` deeply immutable

> 2. The function ``is_frozen(obj)`` -- returns ``True`` if ``obj`` is immutable

Reviewer 3: `isfrozen`

> 3. The function ``is_freezable(obj)`` -- returns ``True`` if all objects reachable from ``obj`` can be frozen

Reviewer 3: `isfreezable`

Making an object *deeply* immutable recursively renders the object
*and all objects it references* immutable. (Just
making the first object immutable is called *shallow*
immutability.)

Deep immutability provides strong guarantees against
unintended modifications, thereby improving correctness, security, and
parallel execution safety.

Immutable objects are managed with reference counting plus cycle
detection just like normal mutable objects, and can be made
immortal. In this PEP, we rely on the GIL to ensure the
correctness of reference counts of immutable objects, but we have
several planned memory management extensions, including support
for atomic reference counting on immutable objects. These are
outlined at the end of this document.

Immutability in action:

.. code-block:: python

  from immutable import freeze, is_frozen

  class Foo:
      pass

  f = Foo()
  g = Foo()
  h = Foo()

  f.f = g
  g.f = h
  h.f = g # cycles are OK!
  del g # Remove local ref to g, so g's RC = 1
  del h # Remove local reg to h, so h's RC = 1

  g.x = "African Swallow" # OK
  freeze(f) # Makes, f, g and h immutable
  g.x = "European Swallow" # Throws an exception "g is immutable"

>   is_frozen(h) # returns True

Reviewer 3: `isfrozen`

flag in the header of ``op`` and signals an error if the immutability
flag is set.

A typical use of this check looks like this:

.. code-block:: c

  if (!Py_CHECKWRITE(op)) {        // perform the check
      PyErr_WriteToImmutable(op);  // raise the error if the check fails
      return NULL;                 // abort the write
  }  
  ... // code that performs the write

Writes are common in the CPython code base and the writes lack a
common "code path" that they pass. To this end, the PEP requires a
``Py_CHECKWRITE`` call to be inserted and there are several places
in the CPython code base that are changed as a consequence of this
PEP. So far we have identified around 70 places in core Python which
needed a ``Py_CHECKWRITE`` check. Modules in the standard library
have required somewhere between 5 and 15 checks per module.

Backward Compatibility
======================

This proposal intends to be fully backward compatible, as no existing Python
code will be affected unless it explicitly calls ``freeze(obj)``.
Frozen objects will raise errors only when mutation is attempted.

Opt-In vs. Opt-Out
------------------

> All pure Python objects can be frozen. However, for types which

Reviewer 3: suggested edit: All pure Python objects can be implicitly frozen, provided all their members and their base classes can be frozen.

Author: I added it in but dropped the word "implicitly" because that was confusing.

are partially or completely implemented in C, support for
immutability requires some work on both exposing objects to
freezing, and to enforce immutability in mutating C-functions.

From a backwards compatibility perspective, an opt-in model keeps
things simple: all existing code keeps working, and only code that
wishes to support immutability needs updating. The downside of the
opt-in model is that a large part of all Python libraries cannot
be (even nominally) frozen.

> This PEP proposes to make support for immutability in C extensions

Reviewer 3: add: and Python wrappers of classes which would otherwise not be freezable

4. **Rejected alternative**: Permit the user to choose between
   alternatives 1) and 3) at use-site. In this case, the
   ``freeze`` function takes an optional 2nd argument ``strict``
   which must either be ``True`` or ``False``. In the first case,
   ``freeze`` behaves as in alternative 1), in the second case,
   it behaves as in alternative 2). We could further track whether
   an object is strictly immutable or not in order to prevent
   non-strictly immutable objects to participate in operations
   which require strictness. This adds additional complexity to
   the implementation, and also for the user.
   
This PEP proposes following alternative 1, where freezing either
succeeds or fails completely.

New Obligations on C Extensions
-------------------------------

Due to the opt-in decision, there are no *obligations* for C
extensions that does not want to add support for immutability. 

Because our implementation builds on information available to the CPython
cycle detector, types defined through C code will support immutability 
"out of the box" as long as they use Python standard types to store
data and uses the built-in functions of these types to modify the data.

> To make its instances freezable, types using C extensions must

Reviewer 3: Just as a note: this is only true if the C extension adds new functionality implemented in C (which isn't always the case). It is a subtle point but worth making. If you construct a C type using freezable metaclasses it will itself be freezable.

> register themselves using ``register_freezable(type)``. To

Reviewer 3: Here is some sample code you could potentially include:

```c
    PyObject *register_freezable = _PyImport_GetModuleAttrString("immutable", "register_freezable");
    if(register_freezable != NULL)
    {
        PyObject* result = PyObject_CallOneArg(register_freezable, (PyObject *)st->Element_Type);
        if(result == NULL){
            goto error;
        }

        Py_DECREF(register_freezable);
    }
```

              return a
          else:
              x = x + 1
              return x

      freeze(foo)
      foo(41)  # OK, returns 42
      foo()    # Throws NotWriteableError

  example2()

This example illustrates two things. The first call to ``foo(41)``
shows that local variables on the frame of a frozen function are
mutable. The second call shows that captured variables are not.
Note that the default value of ``a`` will be frozen when ``foo``
is frozen. Thus, the problem of side-effects on default values
on parameters is avoided.

Frozen function objects that access globals, e.g. through an
explicit call to ``globals()``, will throw an exception when
called.

Implementation Details
======================

1. Add the ``immutable`` module, the ``NotWriteableError`` type, and
   the ``NotFreezable`` type.
2. Add the ``freeze(obj)`` function to the ``immutable`` module and
   ensure that it traverses object references safely, including
   cycle detection, and marks objects appropriately, and backs
   out on failure, without partially freezing object graphs.
3. Add the ``register_freezable(type)`` whitelisting function.

> 4. Add the ``is_freezable(obj)`` function which checks that all

Reviewer 3: isfreezable

   objects reachable from ``obj`` can be frozen.
5. Add the ``register_freezable(type)`` function that is used to
   whitelist types implemented as C extensions, permitting their
   instances to be frozen.

> 6. Add the ``is_frozen(obj)`` function to the ``immutable`` module

Reviewer 3: isfrozen

   that checks whether or not an object is frozen. The status
   is accessible through ``_Py_ISIMMUTABLE`` in the C API and in
   Python code through the ``is_frozen(obj)`` function.
7. Modify object mutation operations (``PyObject_SetAttr``,
   ``PyDict_SetItem``, ``PyList_SetItem``, etc.) to check the 
   flag and raise an error when appropriate.

> 8. Modify mutation operations in modules in the standard library.

Reviewer 3: Changes to the C ABI:

1. `Py_CHECKWRITE`
2. `_Py_IsImmutable`
3. `PyErr_WriteToImmutable`

Changes to the internal API:
1. `_PyType_HasExtensionSlots(PyTypeObject*)` - Determines whether a TypeObject adds novel functionality in C
2. `_PyNotFreezable_Type`
3. `_PyImmutability_Freeze`
4. `_PyImmutability_RegisterFreezable`
5. `_PyImmutability_IsFreezable`

Author: I'll add these as subsections under the list. Sounds OK?

   and write--write races. This alternative was rejected for two main
   reasons:

    1. It is expensive to implement: it needs a read-barrier to
       detect what objects are being read by threads to capture
       read--write races.
    2. While more permissive, the model suffers from non-determinism.
       Data races can be hidden in corner cases that require complex
       logic and/or temporal interactions which can be hard to
       test and reproduce.

A Note on Modularisation
========================
   
While the ``freeze(obj)`` function is available to Python programmers
in the ``immutable`` module, the actual freezing code has to live
inside core Python. This is for three reasons:

1. The core object type needs to be able to freeze just-in-time
   dictionaries created by its accessors when the object itself is
   frozen.
2. The managed buffer type needs to be frozen when the object it
   is created from is frozen.
3. Teardown of strongly connected components of immutable objects
   (see `Simplified Garbage Collection for Immutable Object
   Graphs`_) must be hooked into ``Py_DECREF``.

As such, we implement a function which is not in the limited API

> (and thus not part of the stable C ABI) called ``_Py_Freeze``

Reviewer 3: Note the name change above

`PEP 703`_'s focus is on making the CPython run-time resilient against
data races in Python programs: a poorly synchronized Python program
should not be able to corrupt reference counts, or other parts of
the Python interpreter. The complementary goal pursued by this PEP
is to make it impossible for Python programs to have data races.
Support for deeply immutable data is the first important step
towards this goal.

Reference Implementation
========================

`Available here <https://github.com/mjp41/cpython/pull/51>`_.

There are some discrepancies between this PEP and the reference
implementation, including:

- Freezing a nested function behaves slightly different from
  freezing a top-level function. The latter behaves as outlined
  in this document. The former also blocks changes to the captured
  variable in the enclosing scope because they currently share a
  common cell object.
- The ``is_freezable(obj)`` function is not yet implemented.
- The ``NotFreezable`` type is currently freezable (but inheriting
  from it stops instances of the inheriting class from being frozen). 

Rebasing on Python 3.14
=======================

> **TODO MJP**

Reviewer: ```suggestion
We have found two areas that need to be addressed to integrate this work with "free-threaded Python": data-representation and data-races during freeze.

Data-representation for immutability
------------------------------------

With free-threaded Python the representation of the reference count has been changed.  We could either borrow a bit to represent if an object is frozen, or alternatively, we could use the new `ob_tid` field to have a special value for immutable state.  Using `ob_tid` would allow for standard mutable thread local objects to remain the fast path, and is our preferred alternative.

The extensions use use SCC calculations to detect cycles in immutable graphs, would require additional state.  Repurposing `ob_tid` and `ob_ref_shared` would allow sufficient space for the necessary calculation.

Data-races during freeze
------------------------

We consider the following races

- Freezing some objects concurrently with another thread checking if a graph is immutable.
- Freezing some objects concurrently with another thread mutating those objects.
- Freezing some objects concurrently with another thread freezing those objects.

To address the first race, we need to consider strictness of deep immutability.  We need to ensure that querying an object graph for immutability only says yes if it is deeply immutable. This requires a two step immutable state: immutable but not strict, and then immutable and strict. On a DFS traversal of the object graph items are marked as immutable but not strict on the pre-order step, and then immutable and strict on the post-order step.  To query if a graph is immutable, we will require the "immutable and strict" state.

Handling mutation during freeze can use the mutex added by free-threading. There are some cases where mutation does not require the acquisition of a mutex, which would no longer allowed with this feature.  Freezing would be required to lock the object, marks it as immutable, release the lock, and then read all its fields.

The final case is the most complex detecting parallel freezing of an object graph.  We will consider this an error.  This error can be detected as follows.  If we encounter an object that is "immutable but not strict", then this should be on the path to the current object from the starting point of the freeze.  If this is not the case, then we must be observing another thread freezing an object graph.  The algorithm should back out the pending aspects of freeze, and raise an exception to the user.  This can naturally be integrated with the SCC algorithm. 
```

   object ``o2``. In this scenario, the object graph ``o1 --> o2
   --> o3`` remains mutable and ``freeze(o1)`` raises an exception
   when the object graph traversal encounters ``o2``.
3. **Rejected alternative**: Freeze succeeds by altering the
   graph. In this example removing ``o2`` from the graph or
   swapping out ``o2`` for a placeholder object to be able to
   freeze the graph. This alternative becomes complicated both to
   reason about from a user's perspective, and to implement when
   ``o2`` is referenced multiple times.
4. **Rejected alternative**: Permit the user to choose between
   alternatives 1) and 3) at use-site. In this case, the
   ``freeze`` function takes an optional 2nd argument ``strict``
   which must either be ``True`` or ``False``. In the first case,
   ``freeze`` behaves as in alternative 1), in the second case,
   it behaves as in alternative 2). We could further track whether
   an object is strictly immutable or not in order to prevent
   non-strictly immutable objects to participate in operations
   which require strictness. This adds additional complexity to
   the implementation, and also for the user.
   

> This PEP proposes following alternative 1, where freezing either
> succeeds or fails completely.

Reviewer 2: Shouldn't this say: "or fails partially"? Option 2 would be failing completely.

```suggestion
This PEP proposes following alternative 1, where freezing either
succeeds or fails partially.
```

Or it could be that the number is wrong and this pep actually proposes success or failing completely which would be "2. rejected alternative"

---

I also wonder if we should add a motivation like "[...] fails partially. We believe this is easier to implement and more in the spirit of what the user wants."

But this could, of course, spark discussions

New Obligations on C Extensions
-------------------------------

Due to the opt-in decision, there are no *obligations* for C

> extensions that does not want to add support for immutability. 

Reviewer 2: ```suggestion
extensions that do not want to add support for immutability. 
```

Spelling error?

  ``y.f = 42``. If the object in ``x`` can reach the same object that
  ``y`` points to, then, the assignment will fail. **Mitigation:** To
  facilitate debugging, exceptions due to attempting to mutate immutable
  objects will include information about on what line an object was made
  immutable.
* Class Freezing: Freezing an instance of a user-defined class
  will also freeze its class. Otherwise, sharing an immutable object
  across threads would lead to sharing its *mutable* type object. Thus,
  freezing an object also freezes the type type object of its super
  classes. This means that any metaprogramming or changes to a class
  must happen before a class is frozen. **Mitigation:** A frozen class
  can be extended and its behaviour overridden through normal object-oriented
  means. If neccessary, it is possible to add an option to make a mutable
  copy of immutable objects and classes, which could then be changed.
  Mutable instances of an immutable class can have their classes changed
  to the mutable copy by reassigning ``__class__``.
* Metaclass Freezing: Since class objects have metaclasses,
  freezing a class may propagate upwards through the metaclass

>   hierarchy. This means that the ``Type`` object will be made immutable

Reviewer 2: Above it was called `type` with a lower letter, which, I believe, is the correct way?

```suggestion
  hierarchy. This means that the ``type`` object will be made immutable
```

  def example2():
      x = 0
      def foo(a = False):
          if a:
              a = a + 1 # Note: updating local variables work, even in a frozen function
              return a
          else:
              x = x + 1
              return x

      freeze(foo)
      foo(41)  # OK, returns 42
      foo()    # Throws NotWriteableError

  example2()

This example illustrates two things. The first call to ``foo(41)``
shows that local variables on the frame of a frozen function are
mutable. The second call shows that captured variables are not.
Note that the default value of ``a`` will be frozen when ``foo``
is frozen. Thus, the problem of side-effects on default values
on parameters is avoided.

Frozen function objects that access globals, e.g. through an
explicit call to ``globals()``, will throw an exception when
called.

Implementation Details
======================

1. Add the ``immutable`` module, the ``NotWriteableError`` type, and
   the ``NotFreezable`` type.
2. Add the ``freeze(obj)`` function to the ``immutable`` module and
   ensure that it traverses object references safely, including
   cycle detection, and marks objects appropriately, and backs

>    out on failure, without partially freezing object graphs.

Reviewer 2: Doesn't section "Dealing with Failure During Freezing" suggest taking the first option with partial freezing? :thinking:

Author: ```suggestion
2. Add the ``freeze(obj)`` function to the ``immutable`` module and
   ensure that it traverses object references safely, including
   cycle detection, and marks objects appropriately, and backs
   out on failure, possibly partially freezing the object graph.
```

3. Add the ``register_freezable(type)`` whitelisting function.
4. Add the ``isfreezable(obj)`` function which checks that all
   objects reachable from ``obj`` can be frozen.
5. Add the ``register_freezable(type)`` function that is used to
   whitelist types implemented as C extensions, permitting their

>    instances to be frozen.

Reviewer 2: Are 3. and 5. supposed to have the same function name? Should we remove 3.

Author: ```suggestion
3. Add the ``isfreezable(obj)`` function which checks that all
   objects reachable from ``obj`` can be frozen.
4. Add the ``register_freezable(type)`` function that is used to
   whitelist types implemented as C extensions, permitting their
   instances to be frozen.
```
